File: integrations/browser/oicana-browser/src/Template.ts
import {
  compile_template,
  export_document,
  get_file,
  get_source,
  get_warnings,
  register_template,
  remove_document,
  remove_world,
  set_validate_inputs,
  inputs as wasmInputs,
} from '@oicana/browser-wasm';
import { CompilationMode } from './CompilationMode.js';
import { CompiledDocument } from './CompiledDocument.js';
import { type ExportFormat, Pdf, Png, Svg } from './ExportFormat.js';
import type {
  BlobInputDefinition,
  BlobWithMetadata,
  JsonInputDefinition,
} from './inputs/index.js';
import type { PageRange } from './PageRange.js';

/**
 * A template
 *
 * The zip file is loaded during the instance creation and cached afterward.
 */
export class Template implements Disposable {
  private readonly template: string;
  private lastWarnings: string | undefined;

  /**
   * Register a template with the given template file
   * @param template - the packed Oicana template file
   */
  public constructor(template: Uint8Array);

  /**
   * Register a template with the given template file and inputs
   * @param template - the packed Oicana template file
   * @param jsonInputs for the initial compilation to warm up the cache
   * @param blobInputs for the initial compilation to warm up the cache
   */
  public constructor(
    template: Uint8Array,
    jsonInputs: Map<string, string>,
    blobInputs: Map<string, BlobWithMetadata>,
  );

  /**
   * Register a template with the given template file and inputs
   * @param template - the packed Oicana template file
   * @param jsonInputs for the initial compilation to warm up the cache (defaults to empty map)
   * @param blobInputs for the initial compilation to warm up the cache (defaults to empty map)
   * @param compilationOptions for the initial compilation to warm up the cache (defaults to Development)
   */
  public constructor(
    template: Uint8Array,
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
  ) {
    this.template = crypto.randomUUID();
    for (const blob of blobInputs?.entries() ?? []) {
      if (blob[1].meta === undefined) {
        // Otherwise the FFI layer will fail to pass the blobs over to WASM
        blob[1].meta = {};
      }
    }
    const documentId = register_template(
      this.template,
      template,
      jsonInputs ?? new Map(),
      blobInputs ?? new Map(),
      compilationOptions ?? CompilationMode.Development,
    );
    this.lastWarnings = get_warnings(documentId);
    remove_document(documentId);
  }

  /**
   * Compile the template and export it to a PDF file, without inputs, in
   * production mode.
   */
  public export(): Uint8Array<ArrayBuffer>;

  /**
   * Compile the template with the given inputs and export it to a PDF file in
   * production mode.
   * @param jsonInputs
   * @param blobInputs
   */
  public export(
    jsonInputs: Map<string, string>,
    blobInputs: Map<string, BlobWithMetadata>,
  ): Uint8Array<ArrayBuffer>;

  /**
   * Compile the template with the given inputs and export it in the given format.
   * @param jsonInputs
   * @param blobInputs
   * @param exportFormat
   */
  public export(
    jsonInputs: Map<string, string>,
    blobInputs: Map<string, BlobWithMetadata>,
    exportFormat: ExportFormat,
  ): Uint8Array<ArrayBuffer>;

  /**
   * Compile the template with the given inputs and export it in the given format.
   * @param jsonInputs
   * @param blobInputs
   * @param exportFormat
   * @param compilationOptions
   */
  public export(
    jsonInputs: Map<string, string>,
    blobInputs: Map<string, BlobWithMetadata>,
    exportFormat: ExportFormat,
    compilationOptions: CompilationMode,
  ): Uint8Array<ArrayBuffer>;

  /**
   * Compile the template with the given inputs and export a range of pages
   * @param jsonInputs
   * @param blobInputs
   * @param exportFormat
   * @param compilationOptions
   * @param pages
   */
  public export(
    jsonInputs: Map<string, string>,
    blobInputs: Map<string, BlobWithMetadata>,
    exportFormat: ExportFormat,
    compilationOptions: CompilationMode,
    pages: PageRange,
  ): Uint8Array<ArrayBuffer>;

  /**
   * Compile the template and export it.
   *
   * This is the one-shot path: the compiled document is not kept. To export it
   * several times (multiple formats, page ranges, or individual pages) from a
   * single compilation, use {@link compile} and call `export` on the returned
   * {@link CompiledDocument}.
   * @param jsonInputs - JSON inputs for the template (defaults to empty map)
   * @param blobInputs - Blob inputs for the template (defaults to empty map)
   * @param exportFormat - Export format specification (defaults to PDF)
   * @param compilationOptions - Compilation mode (defaults to Production)
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public export(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    exportFormat?: ExportFormat,
    compilationOptions?: CompilationMode,
    pages?: PageRange,
  ): Uint8Array {
    return this.exportWith(
      exportFormat ?? Pdf,
      jsonInputs,
      blobInputs,
      compilationOptions,
      pages,
    );
  }

  /**
   * Compile the template and export it to PDF in a single call, then free the
   * document.
   * @param jsonInputs - JSON inputs for the template (defaults to empty map)
   * @param blobInputs - Blob inputs for the template (defaults to empty map)
   * @param compilationOptions - Compilation mode (defaults to Production)
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportPdf(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
    pages?: PageRange,
  ): Uint8Array {
    return this.exportWith(
      Pdf,
      jsonInputs,
      blobInputs,
      compilationOptions,
      pages,
    );
  }

  /**
   * Compile the template and export it to PNG in a single call, then free the
   * document.
   * Multiple pages are merged into a single, vertically stacked image.
   * @param jsonInputs - JSON inputs for the template (defaults to empty map)
   * @param blobInputs - Blob inputs for the template (defaults to empty map)
   * @param compilationOptions - Compilation mode (defaults to Production)
   * @param pixelsPerPt - resolution in pixels per point (defaults to 1.0)
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportPng(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
    pixelsPerPt = 1.0,
    pages?: PageRange,
  ): Uint8Array {
    return this.exportWith(
      Png(pixelsPerPt),
      jsonInputs,
      blobInputs,
      compilationOptions,
      pages,
    );
  }

  /**
   * Compile the template and export it to SVG in a single call, then free the
   * document.
   * @param jsonInputs - JSON inputs for the template (defaults to empty map)
   * @param blobInputs - Blob inputs for the template (defaults to empty map)
   * @param compilationOptions - Compilation mode (defaults to Production)
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportSvg(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
    pages?: PageRange,
  ): Uint8Array {
    return this.exportWith(
      Svg,
      jsonInputs,
      blobInputs,
      compilationOptions,
      pages,
    );
  }

  private exportWith(
    format: ExportFormat,
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
    pages?: PageRange,
  ): Uint8Array {
    const documentId = this.compileToDocumentId(
      jsonInputs,
      blobInputs,
      compilationOptions,
    );
    try {
      return export_document(documentId, format, pages);
    } finally {
      remove_document(documentId);
    }
  }

  /**
   * Compile the template.
   *
   * The document is kept in memory so it can be exported one or more times
   * (whole document, a page range, or individual pages) without re-compiling.
   * Call `dispose()` on the returned document (or use `using`) to free it.
   * For a single one-shot export, prefer {@link export}.
   * @param jsonInputs - JSON inputs for the template (defaults to empty map)
   * @param blobInputs - Blob inputs for the template (defaults to empty map)
   * @param compilationOptions - Compilation mode (defaults to Production)
   */
  public compile(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
  ): CompiledDocument {
    const documentId = this.compileToDocumentId(
      jsonInputs,
      blobInputs,
      compilationOptions,
    );
    return new CompiledDocument(documentId);
  }

  private compileToDocumentId(
    jsonInputs?: Map<string, string>,
    blobInputs?: Map<string, BlobWithMetadata>,
    compilationOptions?: CompilationMode,
  ): string {
    for (const blob of blobInputs instanceof Map
      ? (blobInputs?.entries() ?? [])
      : Object.entries<BlobWithMetadata>(blobInputs ?? {})) {
      if (blob[1].meta === undefined) {
        // Otherwise the FFI layer will fail to pass the blobs over to WASM
        blob[1].meta = {};
      }
    }
    const documentId = compile_template(
      this.template,
      jsonInputs ?? new Map(),
      blobInputs ?? new Map(),
      compilationOptions ?? CompilationMode.Production,
    );
    this.lastWarnings = get_warnings(documentId);
    return documentId;
  }

  /**
   * Warnings produced by the most recent compilation (constructor warm-up, or a
   * `compile()` / `export()` call), or `undefined` if there were none.
   */
  public warnings(): string | undefined {
    return this.lastWarnings;
  }

  /**
   * Gather all input definitions of this template
   */
  public inputs(): { inputs: (BlobInputDefinition | JsonInputDefinition)[] } {
    return JSON.parse(wasmInputs(this.template));
  }

  /**
   * Get the string content of a file from the template
   */
  public source(path: string): string {
    return get_source(this.template, path);
  }

  /**
   * Get the raw file from the template
   */
  public file(path: string): Uint8Array {
    return get_file(this.template, path);
  }

  /**
   * Enable or disable JSON schema validation for this template.
   *
   * When enabled (the default), JSON inputs are validated against their schemas
   * before compilation.
   * @param validate - whether to validate inputs against their JSON schemas
   */
  public setValidateInputs(validate: boolean): void {
    set_validate_inputs(this.template, validate);
  }

  /**
   * Release resources associated with this template.
   * After calling dispose(), this template instance should not be used.
   */
  public dispose(): void {
    remove_world(this.template);
  }

  /**
   * Enables use with the `using` keyword for automatic resource cleanup.
   */
  [Symbol.dispose](): void {
    this.dispose();
  }
}

File: integrations/java/oicana/src/main/java/com/oicana/Template.java
package com.oicana;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * A template for generating documents.
 *
 * <p>The template ZIP file is loaded during construction and cached for subsequent compilations.
 * Template instances are thread-safe and can be shared across threads.
 *
 * <p>Use try-with-resources for automatic cleanup:
 * <pre>{@code
 * try (var template = new Template(zipBytes)) {
 *     byte[] pdf = template.export(Map.of("name", "{\"value\": \"World\"}"), Map.of());
 * }
 * }</pre>
 */
public class Template implements AutoCloseable {
    private final String templateId;
    private volatile boolean closed = false;

    /**
     * Register a template with the given template file.
     * The template is compiled once in development mode to warm up the cache.
     *
     * @param templateFile the packed Oicana template ZIP file
     */
    public Template(byte[] templateFile) {
        this(templateFile, Map.of(), Map.of(), CompilationMode.DEVELOPMENT);
    }

    /**
     * Register a template with the given template file and inputs.
     *
     * @param templateFile the packed Oicana template ZIP file
     * @param jsonInputs   JSON inputs for the initial warm-up compilation
     * @param blobInputs   blob inputs for the initial warm-up compilation
     * @param mode         compilation mode for the initial warm-up compilation
     */
    public Template(byte[] templateFile, Map<String, String> jsonInputs,
                    Map<String, BlobInput> blobInputs, CompilationMode mode) {
        this(templateFile, jsonInputs, blobInputs, mode, null);
    }

    /**
     * Register a template with the given template file, inputs, and zip limits.
     *
     * @param templateFile the packed Oicana template ZIP file
     * @param jsonInputs   JSON inputs for the initial warm-up compilation
     * @param blobInputs   blob inputs for the initial warm-up compilation
     * @param mode         compilation mode for the initial warm-up compilation
     * @param limits       limits for reading the template ZIP, or {@code null} for the defaults
     */
    public Template(byte[] templateFile, Map<String, String> jsonInputs,
                    Map<String, BlobInput> blobInputs, CompilationMode mode, ZipLimits limits) {
        this.templateId = UUID.randomUUID().toString();
        String documentId = OicanaNative.registerTemplate(
                this.templateId,
                templateFile,
                jsonInputs,
                convertBlobInputs(blobInputs),
                mode.value,
                maxEntries(limits),
                maxTotalDecompressedBytes(limits)
        );
        OicanaNative.removeDocument(documentId);
    }

    /**
     * Compile the template to PDF without any inputs in production mode.
     *
     * @return the compiled document as a byte array
     */
    public byte[] export() {
        return export(Map.of(), Map.of(), ExportFormat.pdf(), CompilationMode.PRODUCTION);
    }

    /**
     * Compile the template without any inputs using the given export format and compilation mode.
     *
     * @param exportFormat the output format
     * @param mode         the compilation mode
     * @return the compiled document as a byte array
     */
    public byte[] export(ExportFormat exportFormat, CompilationMode mode) {
        return export(Map.of(), Map.of(), exportFormat, mode);
    }

    /**
     * Compile the template to PDF with the given inputs in production mode.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @return the compiled document as a byte array
     */
    public byte[] export(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs) {
        return export(jsonInputs, blobInputs, ExportFormat.pdf(), CompilationMode.PRODUCTION);
    }

    /**
     * Compile the template with the given inputs, export format, and compilation mode.
     *
     * @param jsonInputs   the JSON inputs for the template
     * @param blobInputs   the blob inputs for the template
     * @param exportFormat the output format
     * @param mode         the compilation mode
     * @return the compiled document as a byte array
     */
    public byte[] export(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                          ExportFormat exportFormat, CompilationMode mode) {
        return export(jsonInputs, blobInputs, exportFormat, mode, null);
    }

    /**
     * Compile the template and export a range of pages.
     *
     * @param jsonInputs   the JSON inputs for the template
     * @param blobInputs   the blob inputs for the template
     * @param exportFormat the output format
     * @param mode         the compilation mode
     * @param pages        the 0-based, inclusive page range to export, or {@code null} for the
     *                     whole document
     * @return the compiled document as a byte array
     */
    public byte[] export(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                          ExportFormat exportFormat, CompilationMode mode, PageRange pages) {
        ensureNotClosed();
        String documentId = OicanaNative.compileTemplate(
                this.templateId,
                jsonInputs,
                convertBlobInputs(blobInputs),
                mode.value
        );
        try {
            return OicanaNative.exportDocument(
                    documentId,
                    exportFormat.toJsonString(),
                    pages == null ? null : pages.toJsonString()
            );
        } finally {
            OicanaNative.removeDocument(documentId);
        }
    }

    /**
     * Compile the template to PDF without any inputs in production mode.
     *
     * @return the PDF document as a byte array
     */
    public byte[] exportPdf() {
        return exportPdf(Map.of(), Map.of());
    }

    /**
     * Compile the template to PDF with the given inputs in production mode.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @return the PDF document as a byte array
     */
    public byte[] exportPdf(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs) {
        return exportPdf(jsonInputs, blobInputs, CompilationMode.PRODUCTION, null);
    }

    /**
     * Compile the template and export it to PDF, optionally restricted to a range of pages.
     * Tagging will be automatically turned off when exporting a subset of pages.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @param mode       the compilation mode
     * @param pages      the 0-based, inclusive page range to export, or {@code null} for the
     *                   whole document
     * @return the PDF document as a byte array
     */
    public byte[] exportPdf(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                             CompilationMode mode, PageRange pages) {
        return export(jsonInputs, blobInputs, ExportFormat.pdf(), mode, pages);
    }

    /**
     * Compile the template to PNG without any inputs in production mode, at the default
     * resolution of 1 pixel per point.
     * Multiple pages are merged into a single, vertically stacked image.
     *
     * @return the PNG image as a byte array
     */
    public byte[] exportPng() {
        return exportPng(Map.of(), Map.of(), 1.0f);
    }

    /**
     * Compile the template to PNG without any inputs in production mode.
     * Multiple pages are merged into a single, vertically stacked image.
     *
     * @param pixelsPerPt resolution in pixels per point
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(float pixelsPerPt) {
        return exportPng(Map.of(), Map.of(), pixelsPerPt);
    }

    /**
     * Compile the template to PNG with the given inputs in production mode, at the default
     * resolution of 1 pixel per point.
     * Multiple pages are merged into a single, vertically stacked image.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs) {
        return exportPng(jsonInputs, blobInputs, 1.0f);
    }

    /**
     * Compile the template to PNG with the given inputs in production mode.
     * Multiple pages are merged into a single, vertically stacked image.
     *
     * @param jsonInputs  the JSON inputs for the template
     * @param blobInputs  the blob inputs for the template
     * @param pixelsPerPt resolution in pixels per point
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                             float pixelsPerPt) {
        return exportPng(jsonInputs, blobInputs, CompilationMode.PRODUCTION, pixelsPerPt, null);
    }

    /**
     * Compile the template and export it to PNG, optionally restricted to a range of pages.
     * Multiple pages are merged into a single, vertically stacked image.
     *
     * @param jsonInputs  the JSON inputs for the template
     * @param blobInputs  the blob inputs for the template
     * @param mode        the compilation mode
     * @param pixelsPerPt resolution in pixels per point
     * @param pages       the 0-based, inclusive page range to export, or {@code null} for the
     *                    whole document
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                             CompilationMode mode, float pixelsPerPt, PageRange pages) {
        return export(jsonInputs, blobInputs, ExportFormat.png(pixelsPerPt), mode, pages);
    }

    /**
     * Compile the template to SVG without any inputs in production mode.
     *
     * @return the SVG document as a byte array
     */
    public byte[] exportSvg() {
        return exportSvg(Map.of(), Map.of());
    }

    /**
     * Compile the template to SVG with the given inputs in production mode.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @return the SVG document as a byte array
     */
    public byte[] exportSvg(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs) {
        return exportSvg(jsonInputs, blobInputs, CompilationMode.PRODUCTION, null);
    }

    /**
     * Compile the template and export it to SVG, optionally restricted to a range of pages.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @param mode       the compilation mode
     * @param pages      the 0-based, inclusive page range to export, or {@code null} for the
     *                   whole document
     * @return the SVG document as a byte array
     */
    public byte[] exportSvg(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                             CompilationMode mode, PageRange pages) {
        return export(jsonInputs, blobInputs, ExportFormat.svg(), mode, pages);
    }

    /**
     * Compile the template in production mode without any inputs.
     *
     * @return a handle to the compiled document
     */
    public CompiledDocument compile() {
        return compile(Map.of(), Map.of(), CompilationMode.PRODUCTION);
    }

    /**
     * Compile the template with the given inputs in production mode.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @return a handle to the compiled document
     */
    public CompiledDocument compile(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs) {
        return compile(jsonInputs, blobInputs, CompilationMode.PRODUCTION);
    }

    /**
     * Compile the template.
     *
     * <p>Unlike {@link #export(Map, Map, ExportFormat, CompilationMode)}, the document is kept in
     * memory so it can be exported one or more times without re-compiling. Use try-with-resources
     * or call {@link CompiledDocument#close()} to free it.
     *
     * @param jsonInputs the JSON inputs for the template
     * @param blobInputs the blob inputs for the template
     * @param mode       the compilation mode
     * @return a handle to the compiled document
     */
    public CompiledDocument compile(Map<String, String> jsonInputs, Map<String, BlobInput> blobInputs,
                                    CompilationMode mode) {
        ensureNotClosed();
        String documentId = OicanaNative.compileTemplate(
                this.templateId,
                jsonInputs,
                convertBlobInputs(blobInputs),
                mode.value
        );
        return new CompiledDocument(documentId);
    }

    /**
     * Compile and export a template in a single call without caching.
     * Useful for one-off compilations where template reuse is not needed.
     *
     * @param templateFile the packed Oicana template ZIP file
     * @param jsonInputs   the JSON inputs
     * @param blobInputs   the blob inputs
     * @param exportFormat the output format
     * @param mode         the compilation mode
     * @return the exported document and any compilation warnings
     */
    public static ExportOnceResult exportOnce(byte[] templateFile, Map<String, String> jsonInputs,
                                              Map<String, BlobInput> blobInputs,
                                              ExportFormat exportFormat, CompilationMode mode) {
        return exportOnce(templateFile, jsonInputs, blobInputs, exportFormat, mode, null, null);
    }

    /**
     * Compile and export a template in a single call without caching.
     * Useful for one-off compilations where template reuse is not needed.
     *
     * @param templateFile the packed Oicana template ZIP file
     * @param jsonInputs   the JSON inputs
     * @param blobInputs   the blob inputs
     * @param exportFormat the output format
     * @param mode         the compilation mode
     * @param pages        the 0-based, inclusive page range to export, or {@code null} for the
     *                     whole document
     * @param limits       limits for reading the template ZIP, or {@code null} for the defaults
     * @return the exported document and any compilation warnings
     */
    public static ExportOnceResult exportOnce(byte[] templateFile, Map<String, String> jsonInputs,
                                              Map<String, BlobInput> blobInputs,
                                              ExportFormat exportFormat, CompilationMode mode,
                                              PageRange pages, ZipLimits limits) {
        Object[] result = OicanaNative.exportTemplateOnce(
                templateFile,
                jsonInputs,
                convertBlobInputs(blobInputs),
                mode.value,
                exportFormat.toJsonString(),
                pages == null ? null : pages.toJsonString(),
                maxEntries(limits),
                maxTotalDecompressedBytes(limits)
        );
        return new ExportOnceResult((byte[]) result[0], (String) result[1]);
    }

    /**
     * Get the input definitions for this template as a JSON string.
     *
     * @return JSON string describing the template's input definitions
     */
    public String inputs() {
        ensureNotClosed();
        return OicanaNative.inputs(this.templateId);
    }

    /**
     * Get the source code of a file within the template.
     *
     * @param path the file path within the template
     * @return the source code as a string
     */
    public String source(String path) {
        ensureNotClosed();
        return OicanaNative.getSource(this.templateId, path);
    }

    /**
     * Get the binary content of a file within the template.
     *
     * @param path the file path within the template
     * @return the file content as a byte array
     */
    public byte[] file(String path) {
        ensureNotClosed();
        return OicanaNative.getFile(this.templateId, path);
    }

    /**
     * Enable or disable JSON schema validation for this template.
     *
     * <p>When enabled (the default), JSON inputs are validated against their schemas
     * before compilation.
     *
     * @param validate whether to validate inputs against their JSON schemas
     */
    public void setValidateInputs(boolean validate) {
        ensureNotClosed();
        OicanaNative.setValidateInputs(this.templateId, validate);
    }

    /**
     * Configure automatic cache eviction after each compilation.
     *
     * @param maxAge maximum age threshold. Use -1 to disable eviction, 0 to clear all,
     *               or a positive value to keep entries used within the last n evictions.
     *               Default is 10.
     */
    public static void configureAutomaticCacheEviction(int maxAge) {
        OicanaNative.configureAutomaticCacheEviction(maxAge);
    }

    /**
     * Manually evict the cache with the given age threshold.
     *
     * @param maxAge the age threshold for cache eviction
     */
    public static void evictCache(int maxAge) {
        OicanaNative.evictCache(maxAge);
    }

    /**
     * Release native resources associated with this template.
     * After calling close(), this template instance should not be used.
     */
    @Override
    public void close() {
        if (!closed) {
            closed = true;
            OicanaNative.removeWorld(this.templateId);
        }
    }

    private void ensureNotClosed() {
        if (closed) {
            throw new OicanaException("Template has been closed");
        }
    }

    private static long maxEntries(ZipLimits limits) {
        return limits == null || limits.maxEntries() == null ? -1 : limits.maxEntries();
    }

    private static long maxTotalDecompressedBytes(ZipLimits limits) {
        return limits == null || limits.maxTotalDecompressedBytes() == null
                ? -1
                : limits.maxTotalDecompressedBytes();
    }

    private static Map<String, NativeBlobWithMetadata> convertBlobInputs(Map<String, BlobInput> blobInputs) {
        if (blobInputs == null || blobInputs.isEmpty()) {
            return Map.of();
        }
        Map<String, NativeBlobWithMetadata> result = new HashMap<>(blobInputs.size());
        for (var entry : blobInputs.entrySet()) {
            BlobInput blob = entry.getValue();
            String meta = blob.metadata() == null ? "{}" : toJson(blob.metadata());
            result.put(entry.getKey(), new NativeBlobWithMetadata(blob.data(), meta));
        }
        return result;
    }

    private static String toJson(Map<String, Object> map) {
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (var entry : map.entrySet()) {
            if (!first) sb.append(",");
            first = false;
            sb.append("\"").append(escapeJson(entry.getKey())).append("\":");
            sb.append(valueToJson(entry.getValue()));
        }
        sb.append("}");
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private static String valueToJson(Object value) {
        if (value == null) return "null";
        if (value instanceof String s) return "\"" + escapeJson(s) + "\"";
        if (value instanceof Number n) return n.toString();
        if (value instanceof Boolean b) return b.toString();
        if (value instanceof Map<?, ?> m) return toJson((Map<String, Object>) m);
        if (value instanceof Object[] arr) {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < arr.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(valueToJson(arr[i]));
            }
            sb.append("]");
            return sb.toString();
        }
        if (value instanceof Iterable<?> iter) {
            StringBuilder sb = new StringBuilder("[");
            boolean first = true;
            for (Object item : iter) {
                if (!first) sb.append(",");
                first = false;
                sb.append(valueToJson(item));
            }
            sb.append("]");
            return sb.toString();
        }
        return "\"" + escapeJson(value.toString()) + "\"";
    }

    static String escapeJson(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 8);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '\\' -> sb.append("\\\\");
                case '"' -> sb.append("\\\"");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                case '\b' -> sb.append("\\b");
                case '\f' -> sb.append("\\f");
                default -> {
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
                }
            }
        }
        return sb.toString();
    }
}
